auth/src/user.rs

```diff
@@ -0,0 +1,287 @@
+    http::request::Parts,
+    TypedHeader,
+};
+use rand::distributions::{Alphanumeric, DistString};
+use serde::{Deserialize, Deserializer, Serialize};
+use sqlx::{query, Pool, Row, Sqlite};
+use tracing::{trace, Span};
+
+use crate::{api::RouterState, error::Error};
+
+#[async_trait]
+pub(crate) trait UserManagement {
+    async fn create_user(&self, name: AccountName) -> Result<User, Error>;
+    async fn get_user(&self, name: AccountName) -> Result<User, Error>;
+    async fn get_user_by_key(&self, key: Key) -> Result<User, Error>;
+}
+
+#[derive(Clone)]
+pub(crate) struct UserManager {
+    pub(crate) pool: Pool<Sqlite>,
+}
+
+#[async_trait]
+impl UserManagement for UserManager {
+    async fn create_user(&self, name: AccountName) -> Result<User, Error> {
+        let key = Key::new_random();
+
+        query("INSERT INTO users (account_name, key) VALUES (?1, ?2)")
+            .bind(&name)
+            .bind(&key)
+            .execute(&self.pool)
+            .await?;
+
+        Ok(User::new_with_defaults(name, key))
+    }
+
+    async fn get_user(&self, name: AccountName) -> Result<User, Error> {
+        query("SELECT account_name, key, account_tier FROM users WHERE account_name = ?1")
+            .bind(&name)
+            .fetch_optional(&self.pool)
+            .await?
+            .map(|row| {
+                let permissions = Permissions::builder()
+                    .tier(row.try_get("account_tier").unwrap())
+                    .build();
+
+                User {
+                    name,
+                    key: row.try_get("key").unwrap(),
+                    permissions,
+                }
+            })
+            .ok_or(Error::UserNotFound)
+    }
+
+    async fn get_user_by_key(&self, key: Key) -> Result<User, Error> {
+        query("SELECT account_name, key, account_tier FROM users WHERE key = ?1")
+            .bind(&key)
+            .fetch_optional(&self.pool)
+            .await?
+            .map(|row| {
+                let permissions = Permissions::builder()
+                    .tier(row.try_get("account_tier").unwrap())
+                    .build();
+
+                User {
+                    name: row.try_get("account_name").unwrap(),
+                    key,
+                    permissions,
+                }
+            })
+            .ok_or(Error::UserNotFound)
+    }
+}
+
+#[derive(Clone, Deserialize, PartialEq, Eq, Serialize, Debug)]
+pub struct User {
+    pub name: AccountName,
+    pub key: Key,
+    pub permissions: Permissions,
+}
+
+impl User {
+    #[allow(unused)]
+    pub fn is_admin(&self) -> bool {
+        self.permissions.tier() == &AccountTier::Admin
+    }
+
+    pub fn new_with_defaults(name: AccountName, key: Key) -> Self {
+        Self {
+            name,
+            key,
+            permissions: Permissions::default(),
+        }
+    }
+
+    async fn retrieve_from_key(user_manager: &UserManager, key: Key) -> Result<User, Error> {
```

Comment:
We'll have to invert this pattern so that nothing relies on `UserManager`. Rather everything should use `UserManagement`

Reply:
I ended up removing this method since it was redundant. Does that resolve this, or was there more to it?

Reply:
Nope, making this method redundant was the idea :grin: